libraries/core_libs/network/src/tarcap/packets_handlers/get_pbft_sync_packet_handler.cpp

```diff
@@ -32,38 +32,58 @@ void GetPbftSyncPacketHandler::process(const PacketData &packet_data,
   // Here need PBFT chain size, not synced period since synced blocks has not verified yet.
   const size_t my_chain_size = pbft_chain_->getPbftChainSize();
   size_t blocks_to_transfer = 0;
+  auto will_be_synced = false;
   if (my_chain_size >= height_to_sync) {
-    blocks_to_transfer = std::min(network_sync_level_size_, (my_chain_size - (height_to_sync - 1)));
+    auto total_sync_blocks_size = my_chain_size - height_to_sync + 1;
+    if (total_sync_blocks_size <= network_sync_level_size_) {
+      blocks_to_transfer = total_sync_blocks_size;
+      will_be_synced = true;
+    } else {
+      blocks_to_transfer = network_sync_level_size_;
+    }
+    LOG(log_tr_) << "Will send " << blocks_to_transfer << " PBFT blocks to " << packet_data.from_node_id_;
+  } else {
+    LOG(log_wr_) << "Request syncing period is ahead of own PBFT chain. Peer " << packet_data.from_node_id_
+                 << " want to sync PBFT blocks period from " << height_to_sync << ". But own PBFT chain size is "
+                 << my_chain_size;
+    will_be_synced = true;
   }
-  LOG(log_tr_) << "Will send " << blocks_to_transfer << " PBFT blocks to " << packet_data.from_node_id_;
 
-  sendPbftBlocks(packet_data.from_node_id_, height_to_sync, blocks_to_transfer);
+  sendPbftBlocks(packet_data.from_node_id_, height_to_sync, blocks_to_transfer, will_be_synced);
 }
 
 // api for pbft syncing
 void GetPbftSyncPacketHandler::sendPbftBlocks(dev::p2p::NodeID const &peer_id, size_t height_to_sync,
-                                              size_t blocks_to_transfer) {
+                                              size_t blocks_to_transfer, bool will_be_synced) {
   LOG(log_tr_) << "sendPbftBlocks: peer want to sync from pbft chain height " << height_to_sync
                << ", will send at most " << blocks_to_transfer << " pbft blocks to " << peer_id;
-  uint64_t current_period = height_to_sync;
+  dev::RLPStream s;
+  auto pbft_chain_synced = false;
   if (blocks_to_transfer == 0) {
-    LOG(log_dg_) << "Sending empty PbftSyncPacket to " << peer_id;
-    sealAndSend(peer_id, SubprotocolPacketType::PbftSyncPacket, dev::RLPStream(0));
+    pbft_chain_synced = true;
+    s << pbft_chain_synced;
+    sealAndSend(peer_id, SubprotocolPacketType::PbftSyncPacket, std::move(s));
     return;
   }
 
+  uint64_t current_period = height_to_sync;
   while (current_period < height_to_sync + blocks_to_transfer) {
     bool last_block = (current_period == height_to_sync + blocks_to_transfer - 1);
     auto data = db_->getPeriodDataRaw(current_period);
 
     if (data.size() == 0) {
```

Comment:
I think this should not be possible at all so it could be assert ? Each node node broadcasts its chain size through StatusPacket -> based on that no node should ask for a period that is higher of what the other node has... We should validate this in `GetPbftSyncPacketHandler::process`. If the packet gets through that condition, the only situation when this(we dont have the requested block in db) could happen if there is some serious bug in our code -> we can use assert

Reply:
I agree. Fixed, thanks